File: klox/src/main/kotlin/dev/craigfurman/klox/Resolver.kt
package dev.craigfurman.klox

class Resolver(private val interpreter: Interpreter, private val errorReporter: ErrorReporter) :
    Expression.Visitor<Unit>,
    Stmt.Visitor<Unit> {
    private val scopes = ArrayDeque<MutableMap<String, Boolean>>()
    private var currentFunction = FunctionType.NONE
    private var currentLoop = LoopType.NONE
    private var currentClass = ClassType.NONE

    fun resolve(statements: List<Stmt>) {
        for (stmt in statements) {
            resolve(stmt)
        }
    }

    override fun visitAssignExpr(expr: Expression.Assign) {
        resolve(expr.value)
        resolveLocal(expr, expr.name)
    }

    override fun visitBinaryExpr(expr: Expression.Binary) {
        resolve(expr.left)
        resolve(expr.right)
    }

    override fun visitCallExpr(expr: Expression.Call) {
        resolve(expr.callee)
        for (argument in expr.arguments) {
            resolve(argument)
        }
    }

    override fun visitGetExpr(expr: Expression.Get) = resolve(expr.obj)

    override fun visitGroupingExpr(expr: Expression.Grouping) = resolve(expr.expr)

    override fun visitListExpr(expr: Expression.ListExpr) {
        for (element in expr.elements) {
            resolve(element)
        }
    }

    override fun visitLiteralExpr(expr: Expression.Literal) {
    }

    override fun visitLogicalExpr(expr: Expression.Logical) {
        resolve(expr.left)
        resolve(expr.right)
    }

    override fun visitSetExpr(expr: Expression.SetExpr) {
        resolve(expr.value)
        resolve(expr.obj)
    }

    override fun visitSuperExpr(expr: Expression.Super) {
        TODO("Not yet implemented")
    }

    override fun visitThisExpr(expr: Expression.This) {
        if (currentClass == ClassType.NONE) {
            errorReporter.error(expr.keyword, "Can't use 'this' outside of a class.")
            return
        }
        resolveLocal(expr, expr.keyword)
    }

    override fun visitUnaryExpr(expr: Expression.Unary) {
        resolve(expr.right)
    }

    override fun visitVariableExpr(expr: Expression.Variable) {
        val inOwnInitializer = when (val initialized = scopes.lastOrNull()?.get(expr.name.lexeme)) {
            null -> false
            else -> !initialized
        }
        if (inOwnInitializer) {
            errorReporter.error(expr.name, "Can't read local variable in its own initializer.")
        }
        resolveLocal(expr, expr.name)
    }

    override fun visitBlockStmt(stmt: Stmt.Block) {
        beginScope()
        resolve(stmt.statements)
        endScope()
    }

    override fun visitClassStmt(stmt: Stmt.ClassStmt) {
        val enclosingClass = currentClass
        currentClass = ClassType.CLASS

        declare(stmt.name)
        define(stmt.name)

        beginScope()
        scopes.last()["this"] = true

        for (method in stmt.methods) {
            val declaration =
                if (method.name.lexeme == "init") FunctionType.INITIALIZER else FunctionType.METHOD
            resolveFunction(method, declaration)
        }

        endScope()

        currentClass = enclosingClass
    }

    override fun visitExprStmt(stmt: Stmt.Expr) {
        resolve(stmt.expr)
    }

    override fun visitFunctionStmt(stmt: Stmt.FunctionStmt) {
        declare(stmt.name)
        define(stmt.name)
        resolveFunction(stmt, FunctionType.FUNCTION)
    }

    override fun visitIfStmt(stmt: Stmt.If) {
        resolve(stmt.condition)
        resolve(stmt.thenBranch)
        stmt.elseBranch?.let { resolve(it) }
    }

    override fun visitJumpStmt(stmt: Stmt.Jump) {
        if (currentLoop != LoopType.BREAKABLE) {
            errorReporter.error(stmt.keyword, "Can only break out of loops.")
        }
    }

    override fun visitPrintStmt(stmt: Stmt.Print) {
        resolve(stmt.expr)
    }

    override fun visitReturnStmt(stmt: Stmt.Return) {
        if (currentFunction == FunctionType.NONE) {
            errorReporter.error(stmt.keyword, "Can't return from top-level code.")
        }
        stmt.value?.let {
            // Allow early returns from initializers, just not return values
            if (currentFunction == FunctionType.INITIALIZER) {
                errorReporter.error(stmt.keyword, "Can't return a value from an initializer.")
            }
            resolve(it)
        }
    }

    override fun visitVarStmt(stmt: Stmt.Var) {
        declare(stmt.name)
        stmt.initializer?.let { resolve(it) }
        define(stmt.name)
    }

    override fun visitWhileStmt(stmt: Stmt.While) {
        val enclosingLoop = currentLoop
        currentLoop = LoopType.BREAKABLE

        resolve(stmt.condition)
        resolve(stmt.body)

        currentLoop = enclosingLoop
    }

    private fun resolve(expr: Expression) = visit(expr)
    private fun resolve(stmt: Stmt) = visit(stmt)

    private fun beginScope() = scopes.addLast(HashMap())
    private fun endScope() = scopes.removeLast()

    private fun declare(name: Token) {
        val scope = scopes.lastOrNull() ?: return
        if (scope.containsKey(name.lexeme)) {
            errorReporter.error(name, "Already a variable with this name in this scope.")
        }
        scope[name.lexeme] = false
    }

    private fun define(name: Token) {
        val scope = scopes.lastOrNull() ?: return
        scope[name.lexeme] = true
    }

    private fun resolveLocal(expr: Expression, name: Token) {
        for (i in scopes.size - 1 downTo 0) {
            if (scopes[i].containsKey(name.lexeme)) {
                interpreter.resolve(expr, scopes.size - 1 - i)
                return
            }
        }
    }

    private fun resolveFunction(function: Stmt.FunctionStmt, funcType: FunctionType) {
        val enclosingFunction = currentFunction
        currentFunction = funcType

        beginScope()
        for (param in function.params) {
            declare(param)
            define(param)
        }
        resolve(function.body)
        endScope()

        currentFunction = enclosingFunction
    }

    private enum class FunctionType {
        NONE, FUNCTION, INITIALIZER, METHOD,
    }

    private enum class LoopType {
        NONE, BREAKABLE,
    }

    private enum class ClassType {
        NONE, CLASS,
    }
}

File: golox/lox/resolver.go
package lox

type FunctionType string
type ClassType string

const (
	FuncTypeNone   FunctionType = "NONE"
	FuncTypeFunc   FunctionType = "FUNCTION"
	FuncTypeInit   FunctionType = "INITIALIZER"
	FuncTypeMethod FunctionType = "METHOD"

	ClassTypeNone  ClassType = "NONE"
	ClassTypeClass ClassType = "CLASS"
	ClassTypeSub   ClassType = "SUBCLASS"
)

type Resolver struct {
	interpreter     *Interpreter
	scopes          *Stack[map[string]bool]
	currentFunction FunctionType
	currentClass    ClassType
}

func NewResolver(interpreter *Interpreter) *Resolver {
	return &Resolver{
		interpreter:     interpreter,
		scopes:          &Stack[map[string]bool]{},
		currentFunction: FuncTypeNone,
		currentClass:    ClassTypeNone,
	}
}

func (r *Resolver) resolve(statements []Stmt) {
	for _, statement := range statements {
		must(statement.Accept(r))
	}
}

func (r *Resolver) VisitAssignExpr(expr *AssignExpr) (any, error) {
	_, err := expr.expr.Accept(r)
	must(err)
	r.resolveLocal(expr, expr.name)
	return nil, nil
}

func (r *Resolver) VisitBinaryExpr(expr *BinaryExpr) (any, error) {
	_, err := expr.left.Accept(r)
	must(err)
	_, err = expr.right.Accept(r)
	must(err)
	return nil, nil
}

func (r *Resolver) VisitCallExpr(expr *CallExpr) (any, error) {
	_, err := expr.callee.Accept(r)
	must(err)
	for _, arg := range expr.arguments {
		_, err := arg.Accept(r)
		must(err)
	}
	return nil, nil
}

func (r *Resolver) VisitGetExpr(expr *GetExpr) (any, error) {
	return expr.obj.Accept(r)
}

func (r *Resolver) VisitGroupingExpr(expr *GroupingExpr) (any, error) {
	return expr.expr.Accept(r)
}

func (r *Resolver) VisitLiteralExpr(expr *LiteralExpr) (any, error) {
	return nil, nil
}

func (r *Resolver) VisitLogicalExpr(expr *LogicalExpr) (any, error) {
	_, err := expr.left.Accept(r)
	must(err)
	_, err = expr.right.Accept(r)
	must(err)
	return nil, nil
}

func (r *Resolver) VisitSetExpr(expr *SetExpr) (any, error) {
	_, err := expr.obj.Accept(r)
	must(err)
	return expr.value.Accept(r)
}

func (r *Resolver) VisitThisExpr(expr *ThisExpr) (any, error) {
	if r.currentClass == ClassTypeNone {
		tokenError(expr.keyword, "Can't use 'this' outside of a class.")
	}
	r.resolveLocal(expr, expr.keyword)
	return nil, nil
}

func (r *Resolver) VisitUnaryExpr(expr *UnaryExpr) (any, error) {
	return expr.right.Accept(r)
}

func (r *Resolver) VisitVarExpr(expr *VarExpr) (any, error) {
	if r.scopes.IsEmpty() {
		return nil, nil
	}
	initialized, ok := r.scopes.Peek()[expr.name.lexeme]
	if ok && !initialized {
		tokenError(expr.name, "Can't read local variable in its own initializer.")
	}
	r.resolveLocal(expr, expr.name)
	return nil, nil
}

func (r *Resolver) VisitBlockStmt(stmt BlockStmt) error {
	r.beginScope()
	r.resolve(stmt.statements)
	r.endScope()
	return nil
}

func (r *Resolver) VisitClassStmt(stmt ClassStmt) error {
	enclosing := r.currentClass
	r.currentClass = ClassTypeClass

	r.declare(stmt.name)
	r.define(stmt.name)

	r.beginScope()
	r.scopes.Peek()["this"] = true

	for _, method := range stmt.methods {
		kind := FuncTypeMethod
		r.resolveFunction(method, kind)
	}

	r.endScope()

	r.currentClass = enclosing
	return nil
}

func (r *Resolver) VisitExprStmt(stmt ExprStmt) error {
	_, err := stmt.expr.Accept(r)
	return err
}

func (r *Resolver) VisitFuncStmt(stmt FuncStmt) error {
	r.declare(stmt.name)
	r.define(stmt.name)
	r.resolveFunction(stmt, FuncTypeFunc)
	return nil
}

func (r *Resolver) VisitIfStmt(stmt IfStmt) error {
	_, err := stmt.condition.Accept(r)
	must(err)
	must(stmt.thenBr.Accept(r))
	if stmt.elseBr != nil {
		must(stmt.elseBr.Accept(r))
	}
	return nil
}

func (r *Resolver) VisitPrintStmt(stmt PrintStmt) error {
	_, err := stmt.expr.Accept(r)
	return err
}

func (r *Resolver) VisitReturnStmt(stmt ReturnStmt) error {
	if r.currentFunction == FuncTypeNone {
		tokenError(stmt.keyword, "Can't return from top-level code.")
	}
	if stmt.value != nil {
		_, err := stmt.value.Accept(r)
		must(err)
	}
	return nil
}

func (r *Resolver) VisitVarStmt(stmt VarStmt) error {
	r.declare(stmt.name)
	if stmt.initializer != nil {
		_, err := stmt.initializer.Accept(r)
		must(err)
	}
	r.define(stmt.name)
	return nil
}

func (r *Resolver) VisitWhileStmt(stmt WhileStmt) error {
	_, err := stmt.condition.Accept(r)
	must(err)
	return stmt.body.Accept(r)
}

func (r *Resolver) beginScope() {
	r.scopes.Push(map[string]bool{})
}

func (r *Resolver) endScope() {
	r.scopes.Pop()
}

func (r *Resolver) declare(name Token) {
	// The resolver only operates on non-global scopes. The interpreter
	// special-cases the global scope, and will search it if no local scope
	// resolves a given name.
	if r.scopes.IsEmpty() {
		return
	}

	if _, ok := r.scopes.Peek()[name.lexeme]; ok {
		tokenError(name, "Already a variable with this name in this scope.")
	}
	r.scopes.Peek()[name.lexeme] = false
}

func (r *Resolver) define(name Token) {
	if r.scopes.IsEmpty() {
		return
	}
	r.scopes.Peek()[name.lexeme] = true
}

func (r *Resolver) resolveLocal(expr Expr, name Token) {
	for i := r.scopes.Size() - 1; i >= 0; i-- {
		if _, ok := r.scopes.PeekI(i)[name.lexeme]; ok {
			r.interpreter.resolve(expr, r.scopes.Size()-1-i)
			return
		}
	}
}

func (r *Resolver) resolveFunction(fn FuncStmt, kind FunctionType) {
	enclosing := r.currentFunction
	r.currentFunction = kind

	r.beginScope()
	for _, param := range fn.params {
		r.declare(param)
		r.define(param)
	}
	r.resolve(fn.body)
	r.endScope()

	r.currentFunction = enclosing
}
